pkg/httpsig/verify.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package httpsig
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// CreateSignatureInput as per https://tools.ietf.org/html/draft-ietf-httpbis-message-signatures-00#section-3.2.2.
+func CreateSignatureInput(params *SignatureHeaderParams, r *http.Request) (string, error) {
+	err := ensureConsistentSignatureInput(params)
+	if err != nil {
+		return "", fmt.Errorf("invalid signature header params: %w", err)
+	}
+
+	var input []string
+
+	coveredContent := params.Headers
+
+	for i := range coveredContent {
+		var (
+			canonicalValue string
+			err            error
+		)
+
+		contentIdentifier := coveredContent[i]
+
+		canonicalizeNonHTTPHeader, nonHTTP := nonHTTPHeaderContentIdentifierCanonicalizers()[contentIdentifier]
+		switch nonHTTP {
+		case true:
+			canonicalValue = canonicalizeNonHTTPHeader(params, r)
+		default:
+			canonicalValue, err = canonicalizeHTTPHeaderValue(contentIdentifier, r)
+			if err != nil {
+				return "", fmt.Errorf("failed to canonicalize http header %s: %w", contentIdentifier, err)
+			}
+		}
+
+		// An identifier's entry is a US-ASCII string consisting of the
+		// lowercased identifier followed with a colon "":"", a space "" "", and
+		// the identifier's canonicalized value
+		entry := fmt.Sprintf("%s: %s", strings.ToLower(contentIdentifier), canonicalValue)
+		input = append(input, entry)
+	}
+
+	// the signer concatenates together
+	// entries for each identifier in the signature's Covered Content in the
+	// order it occurs in the list, with each entry separated by a newline
+	// ""\n"".
```

Comment:
(Same question everywhere there's a double quote in the comment)

Reply:
Copied as-is from the ietf draft.